tests/for-each-process/src/lib.rs

```diff
@@ -0,0 +1,54 @@
+#![no_std]
+
+use linux_kernel_module::{self, cstr, rcu, sched};
+
+struct PsAuxFile;
+
+impl linux_kernel_module::file_operations::FileOperations for PsAuxFile {
+    const VTABLE: linux_kernel_module::file_operations::FileOperationsVtable =
+        linux_kernel_module::file_operations::FileOperationsVtable::builder::<Self>()
+            .read()
+            .build();
+
+    fn open() -> linux_kernel_module::KernelResult<Self> {
+        Ok(PsAuxFile)
+    }
+}
+
+impl linux_kernel_module::file_operations::Read for PsAuxFile {
+    fn read(
+        &self,
+        _file: &linux_kernel_module::file_operations::File,
+        buf: &mut linux_kernel_module::user_ptr::UserSlicePtrWriter,
+        _offset: u64,
+    ) -> linux_kernel_module::KernelResult<()> {
+        let g = rcu::RcuReadGuard::new();
+        for mut p in sched::each_process(&g) {
+            buf.write(&p.comm())?;
```

Comment:
Writing to a user pointer can sleep, right, so this is incorrect, since you may sleep during an RCU read-side critical section?

Reply:
Yeah I think you're right. How would you do this in C? Make a local variable, fill it in, and then copy_to_user after rcu_read_unlock?

I can do that, I'm not sure how to get Rust to catch it, and I'm curious if Sparse would have caught it.

Reply:
Note that this is tricky, since in the kernel we usually want to avoid dynamically-sized allocation if we can (_especially_ with lengths under user control), so copying the entire output into memory is frowned upon, but also, once we `rcu_read_unlock`, we can't guarantee our `struct task_struct *` linked list doesn't change out from under us.

In practice, something like this would probably be built on the [`seq_file`](https://lwn.net/Articles/22355/) abstraction. `seq_file` users have a "position" abstraction that they must use to track progress through their underlying data source, and in practice, the `seq_file` infrastructure implements a single-page bounce buffer that callers write into (one page at a time), and then `seq_file` copies that back to userspace.

But wait, you might ask, how do you use that position abstraction to keep your position in the `task_struct` linked list across each page of data? Do you have to `get_task_struct` the current finger into the list?

Well, no, and in fact that would be illegal, since it would allow userspace to construct a long-lived reference to a task by `open`ing your file, advancing through it with small reads until it got to the point you wanted, and then just sitting on the `fd`. By sending that `fd` to another process you could construct a circular list of task references, which is no go.

So instead, you would just keep your cursor as an integer "index into the task list", and, each time you were called upon to return another page of data, you would start with `init_task` and advance `N` tasks forward and resume printing.

This is, you may note, both (a) quadratic, and (b) racy.

And yes, this _is_ how almost every file in `/proc` works; If you read (e.g.) `/proc/mounts` and it takes more than a single `read()` call to return the full contents, you can _miss_ mount entries, even ones that existed continually for the entire `open/read/read/close` cycle, if there are concurrent mounts and unmounts.

[This was a fun day at Ksplice](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=593516)

Reply:
I switched to printk, which one hopes is safe from RCU context. (Our `println!` is heap-free.)

Reply:
Yeah, `printk` is safe in any context, even really hostile ones like NMIs or `stop_machine`